src/content/services/monetizationLinkManager.ts

```diff
@@ -0,0 +1,498 @@
+                throw new Error('Could not find details for monetization node');
+              }
+              if (isDisabled) {
+                stopMonetizationPayload.push({
+                  requestId: details.requestId,
+                  intent: 'disable',
+                });
+              } else {
+                startMonetizationPayload.push(details);
+              }
+            } catch {
+              const payloadEntry = await this.checkLink(target);
+              if (payloadEntry) {
+                this.monetizationLinks.set(target, payloadEntry.details);
+                startMonetizationPayload.push(payloadEntry.details);
+              }
+            }
+
+            handledTags.add(target);
+          }
+        } else if (
+          record.type === 'attributes' &&
+          record.attributeName === 'href' &&
+          target instanceof HTMLLinkElement &&
+          target.href !== record.oldValue
+        ) {
+          stopMonetizationPayload.push(this.onRemovedLink(target));
+          const payloadEntry = await this.checkLink(target);
+          if (payloadEntry) {
+            startMonetizationPayload.push(payloadEntry.details);
+          }
+          handledTags.add(target);
+        }
+      }
+    }
+
+    await this.sendStopMonetization(stopMonetizationPayload);
+    void this.sendStartMonetization(startMonetizationPayload);
+  }
+
+  private async onAddedNode(node: Node) {
+    if (node instanceof HTMLElement) {
+      this.dispatchOnMonetizationAttrChangedEvent(node);
+    }
+
+    if (node instanceof HTMLLinkElement) {
+      return await this.onAddedLink(node);
+    }
+    return null;
+  }
+
+  private async onAddedLink(link: HTMLLinkElement) {
+    this.observeLinkAttrs(link);
+    const res = await this.checkLink(link);
+    if (res) {
+      this.monetizationLinks.set(link, res.details);
+    }
+    return res;
+  }
+
+  private onRemovedLink(link: HTMLLinkElement): StopMonetizationPayloadEntry {
+    const details = this.monetizationLinks.get(link);
+    if (!details) {
+      throw new Error(
+        'Could not find details for monetization node ' +
+          // node is removed, so the reference can not be displayed
+          link.outerHTML.slice(0, 200),
+      );
+    }
+
+    this.monetizationLinks.delete(link);
+
+    return { requestId: details.requestId, intent: 'remove' };
+  }
+}
+
+function isDocumentReady(doc: Document) {
+  return (
+    (doc.readyState === 'interactive' || doc.readyState === 'complete') &&
+    doc.visibilityState === 'visible'
+  );
+}
+
+function getMonetizationLinkTags(
+  document: Document,
+  isTopFrame: boolean,
+): HTMLLinkElement[] {
+  if (isTopFrame) {
+    return Array.from(
+      document.querySelectorAll<HTMLLinkElement>('link[rel="monetization"]'),
+    );
+  } else {
+    const monetizationTag = document.querySelector<HTMLLinkElement>(
+      'head link[rel="monetization"]',
+    );
+    return monetizationTag ? [monetizationTag] : [];
+  }
+}
```

Comment:
Let's make this a method of the `LinkManager` class.

Reply:
@raducristianpopa Just in case, also rename class from `MonetizationLinkManager` to `LinkManager`? Or too ambiguous?

Reply:
`MonetizationLinkManager` sounds fine to me.

Reply:
@raducristianpopa I think I'll export these functions from this file, and not add them to the class, so we can unit test them. WDYT?

Reply:
I don't think that this particular function needs testing. It's mostly a query selector.